File: typescript/samples/3-ecommerce/server/products.api.ts
import { Controller, GET } from 'plat'
import type { RouteContext } from 'plat'
import type { Product, ListProductsInput, ListProductsOutput } from '../shared/types'

const products: Map<number, Product> = new Map([
  [1, { id: 1, name: 'Laptop', description: 'High-performance laptop', price: 1299.99, category: 'Electronics', inStock: true, quantity: 5 }],
  [2, { id: 2, name: 'Mouse', description: 'Wireless mouse', price: 29.99, category: 'Accessories', inStock: true, quantity: 50 }],
  [3, { id: 3, name: 'Keyboard', description: 'Mechanical keyboard', price: 149.99, category: 'Accessories', inStock: true, quantity: 20 }],
  [4, { id: 4, name: 'Monitor', description: '4K monitor', price: 399.99, category: 'Electronics', inStock: false, quantity: 0 }],
])

@Controller()
export class ProductsApi {
  @GET()
  async listProducts(input: ListProductsInput = {}, ctx: RouteContext): Promise<ListProductsOutput> {
    const { category, inStock, limit = 10, offset = 0 } = input
    let filtered = Array.from(products.values())

    if (category) filtered = filtered.filter(p => p.category === category)
    if (inStock !== undefined) filtered = filtered.filter(p => p.inStock === inStock)

    return {
      products: filtered.slice(offset, offset + limit),
      total: filtered.length,
    }
  }

  @GET()
  async getProduct(input: { id: number }, ctx: RouteContext): Promise<Product> {
    const product = products.get(input.id)
    if (!product) throw new Error(`Product ${input.id} not found`)
    return product
  }

  @GET()
  async searchProducts(input: { q: string; limit?: number }, ctx: RouteContext) {
    const { q, limit = 10 } = input
    const query = q.toLowerCase()
    const results = Array.from(products.values())
      .filter(p => p.name.toLowerCase().includes(query) || p.description.toLowerCase().includes(query))
      .slice(0, limit)
    return { products: results, total: results.length }
  }
}

File: python/samples/basic/products.api.py
from __future__ import annotations

from plat import Controller, GET, POST, RouteContext

from samples.basic.models import (
    CreateProductInput,
    CreateProductOutput,
    ListProductsInput,
    ListProductsOutput,
    Product,
)


_PRODUCTS = [
    Product(id="p1", name="Apple", price=1.25),
    Product(id="p2", name="Pear", price=1.5),
    Product(id="p3", name="Orange", price=1.75),
]


@Controller("products", {"tag": "Products"})
class ProductsApi:
    @GET()
    def listProducts(
        self,
        input: ListProductsInput,
        ctx: RouteContext,
    ) -> ListProductsOutput:
        items = _PRODUCTS
        if input.q:
            needle = input.q.lower()
            items = [item for item in items if needle in item.name.lower()]

        return ListProductsOutput(
            items=items[: input.limit],
            total=len(items),
            q=input.q,
        )

    @POST({"auth": "public"})
    def createProduct(
        self,
        input: CreateProductInput,
        ctx: RouteContext,
    ) -> CreateProductOutput:
        product = CreateProductOutput(
            id=f"p{len(_PRODUCTS) + 1}",
            name=input.name,
            price=input.price,
        )
        _PRODUCTS.append(Product.model_validate(product))
        return product
